Usability/Keyboard: Add Ctrl+Shift+F shortcut to search the main tracks database

cr7pt0gr4ph7 opened the pull request:
The new shortcut `Ctrl+Shift+F` (`Search globally`[^1]) complements the existing shortcut `Ctrl+F` (`Search in current location`).

The use case for this is pretty simple (and motivated by me teaching dance classes) and probably pretty common elsewhere, too:

> After digging around in my playlists & crates, leaving the computer to teach, and then coming back to it, I want to be able to quickly search for a specific song without having to first pick up the mouse or fiddle around with the trackpad.

[^1]: Ok, it is only "globally" as long as all tracks are present in the main Mixxx database, and not contained in external libraries like iTunes or Rekordbox.[^2]

[^2]: As a later improvement, one could extend `Ctrl+Shift+F` to, either:
a. Actually initiate a global search over all available libraries, or:
b. Make default library to which `Ctrl+Shift+F` jumps configurable in the preferences, or:
c. Provide different shortcuts for the different external libraries.

cr7pt0gr4ph7 committed:
Library: Add LibraryFeature::selectAndActivate()

ronso0 commented:
Note that there is already `Library::searchTracksInCollection` (emitted when clicking actions in `WSearchRelatedTracksMenu`), which (IMHO) allows a much simpler implementation.

cr7pt0gr4ph7 commented:
@ronso0 Thanks, I will have a look! I'll have to place the shortcut handling somewehere else, though, so it also works when the search widget is disabled (this behavioris inherent to `QShortcut`). I'd use a custom key binding in `KeyboardEventFilter`, but then it'd only work when Keyboard shortcuts are enabled.

Sidenote: There seem to be at at least three ways to define keyboard shortcut actions and associated key sequences right now:
1. via `eventFilter`/`keyPressEvent`
2. via `QShortcut`/`QAction`
3. via `KeyboardEventFilter`/`*.kbd.cfg`/`ControlObject` (only active wjen

Option 3 is only active when keyboard shortcuts are enabled in the options, but when it is, it (mostly) overrides the other two.
Option 2 is somewhat integrated with option 3, but it's all bit convoluted.

Maybe it makes sense to integrate the "(Custom) keyboard shortcuts disabled" codepath into `KeyboardEventFilter`, too?

ronso0 commented:
How about adding "Search in Tracks" to the **Library** menu?
And add the shortcut hint to the searchbar tooltip?

cr7pt0gr4ph7 committed:
Library: Add Ctrl+Shift+F shortcut to search the internal database

cr7pt0gr4ph7 committed:
WSearchLineEdit: Update shortcut presentation in the search box tooltip

...so it uses the same format as the rest of the application.

cr7pt0gr4ph7 committed:
Library: Add focusReason parameter to LibraryControl::setLibraryFocus

cr7pt0gr4ph7 committed:
WSearchLineEdit: Correctly detect autocomplete vs. automatic "select all"  on Enter/Return

When the current text selection is due to the automatic selectAll()
when focusing the text field, a subsequent Key_Enter should switch to the
library view. Otherwise, it is due to an autocompletion and should trigger
the search first.

cr7pt0gr4ph7 committed:
WSearchLineEdit: Fix: Remove unsupported Ctrl+Backspace shortcut from tooltip

The Ctrl+Backspace shortcut is a standard shortcut handled by QLineEdit
and deletes the current word, but doesn't actually clear the search box
if more than one word has been typed. This is different from the function
of the clear button, which clears the whole search box.

cr7pt0gr4ph7 committed:
Library: Make search shortcuts configurable via keymap

cr7pt0gr4ph7 committed:
Library: Fix: setLibraryFocus should work even if the window is not currently focused

This scenario occurs e.g. when calling setLibraryFocus from a menu action.
The popup window has already been closed when QAction::triggered occurs,
but the main window has not yet gotten its focus back.

As per the documentation of QWidget::setFocus(), when the containing
window is not active, the focus change request is delayed until the
window becomes active - which is exactly what we want in our case.

cr7pt0gr4ph7 committed:
WSearchLineEdit: Order the shortcuts list in the tooltip by importance

cr7pt0gr4ph7 committed:
WSearchLineEdit: Fix: Add missing whitespace in tooltip text

cr7pt0gr4ph7 committed:
WSearchLineEdit: Improve description text in search box tooltip

cr7pt0gr4ph7 committed:
WSearchLineEdit: Add Ctrl+Return as an alternative to Esc

cr7pt0gr4ph7 committed:
WSearchLineEdit: Use consistent wording ("jump to" vs. "focus") in tooltip

ronso0 commented on src/widget/wsearchlineedit.cpp:
```diff
@@ -809,3 +816,10 @@ void WSearchLineEdit::slotSetFont(const QFont& font) {
     return lineEdit()->hasSelectedText();
 }
+
+bool WSearchLineEdit::hasCompletionAvailable() const {
+    QCompleter* pCompleter = completer();
+    return pCompleter && hasSelectedText() &&
+            lineEdit()->text().startsWith(pCompleter->completionPrefix()) &&
+            pCompleter->completionPrefix().size() == lineEdit()->cursorPosition();
+}
```
```suggestion
bool WSearchLineEdit::hasCompletionAvailable(QString* completionPrefix) const {
    QCompleter* pCompleter = completer();
    if (pCompleter && hasSelectedText() &&
            lineEdit()->text().startsWith(pCompleter->completionPrefix()) &&
            pCompleter->completionPrefix().size() == lineEdit()->cursorPosition()) {
            if (completionPrefix) {
                completionPrefix = pCompleter->completionPrefix();
            }
            return true;
        }
}
```

ronso0 commented on src/widget/wsearchlineedit.cpp:
```diff
@@ -217,31 +211,35 @@ void WSearchLineEdit::setup(const QDomNode& node, const SkinContext& context) {
+            tr("Clear the search bar input field"));
+}
 
+void WSearchLineEdit::setupToolTip(const QString& searchInCurrentViewShortcut,
```
this is not called?

cr7pt0gr4ph7 replied:
Good catch! The changes to LegacySkinParser got lost during cherry-picking the changes for this PR... should be fixed now.

ronso0 reviewed (changes requested):
Thank you for this feature. I did a brief test and everything works as expected.

Here's a first review.

cr7pt0gr4ph7 committed:
WSearchLineEdit: Fix: Invoke WSearchLineEdit::setupTooltip in LegacySkinParser

cr7pt0gr4ph7 committed:
WSearchLineEdit: Review: Rename WSearchLineEdit::handleSetFocus to setFocus

cr7pt0gr4ph7 committed:
WSearchLineEdit: Review: Simplify WSearchLineEdit::getSearchText

cr7pt0gr4ph7 commented:
@ronso0 Thanks for the quick review! 
 requested changes were integrated.

cr7pt0gr4ph7 commented:
@ronso0 Just a friendly ping. Can we get this merged?

cr7pt0gr4ph7 commented:
@daschuer

cr7pt0gr4ph7 commented:
@daschuer or anybody else: Can I do anything to move this PR forward / get it merged? It's been one month...

As for relevance/usability, this is now a main part of my current workflow when using Mixxx, so I think it'll probably be of use to other people, too.

_**PS:** If you want to quickly try it out without having to build it yourself, you can use the `io.github.cr7pt0gr4ph7.Mixxx` Flatpak published at [`cr7pt0gr4ph7/mixxx-flatpak`](https://github.com/cr7pt0gr4ph7/mixxx-flatpak) that includes this and a few other changes. Parallel installation is possible, i.e. any existing official or Flatpak installation of Mixxx will not be affected:_
```
sudo flatpak remote-add --if-not-exists mixxx-flatpak https://cr7pt0gr4ph7.github.io/mixxx-flatpak/default.flatpakrepo
sudo flatpak install io.github.cr7pt0gr4ph7.Mixxx
```

cr7pt0gr4ph7 committed:
Merge branch 'main' into search-in-tracks-shortcut

ronso0 commented:
Sorry for letting this wait.
Unfortunately there is a backlog of open PRs (features, bugfixes & improv) and we're a bit hsort on reviewers currently.
Furthermore, 2.5 bugfixes have priority atm to get a 2.5.1 release out.
I have bookmarked this PR and I intend to get this feature merged in 2.6-alpha, or latest early in beta phase :+1:

acolombier commented on res/keyboard/fr_FR.kbd.cfg:
```diff
@@ -18,9 +18,7 @@ LoadSelectedTrackAndPlay p
-
-[QuickEffectRack1_[Channel1]]
-enabled (
+LibraryMen
```
Is this expected to be incomplete?

cr7pt0gr4ph7 replied:
Ermmmm... no 😅

acolombier reviewed (approved):
The code LGTM! Haven't had a chance to test the feature yet

acolombier reviewed (commented):
Tested and works as expected! @ronso0 were your change requests addressed?

acolombier commented:
Friendly ping @ronso0 :)

ronso0 commented:
Just a note before I take a final look:
we need to document the new feature here https://manual.mixxx.org/2.5/en/chapters/appendix/keyboard_mapping_table#keyboard-mapping-table
(and add anchors for each section, probably simplify use/move the left columns' titles.

ronso0 commented:
LGTM, will do a manual test asap.

Meanwhile, do you mind squashing the "accidental insertion/removal" commits into the kbd.cfg commit?
In case this causes trouble with the merge of main, just leave as is.

cr7pt0gr4ph7 commented:
> Meanwhile, do you mind squashing the "accidental insertion/removal" commits into the kbd.cfg commit? In case this causes trouble with the merge of main, just leave as is.

Done :+1:

daschuer commented:
@ronso0 Can this be merged to 2.6-beta?

ronso0 commented on src/widget/wmainmenubar.cpp:
```diff
@@ -168,6 +168,34 @@ void WMainMenuBar::initialize() {
+    connect(pSearchHere, &QAction::triggered, this, &WMainMenuBar::searchInCurrentView);
+    pLibraryMenu->addAction(pSearchHere);
+
+    QString searchAllTitle = tr("Search in Tracks Library...");
```
Let's use "Tracks view" so it's consistent with "Current view" and the searchbar tooltip

ronso0 reviewed (approved):
Thank you, LGTM and works as expected!

ronso0 commented:
> @ronso0 Can this be merged to 2.6-beta?

Yes.
It also fixes an (searchbar) issue with cursor position reset in main (didn't check 2.5, yet) caused by saveSearch().

ronso0 commented:
Whooopsy, totally forgot to press merge.

Thank you very mucho @cr7pt0gr4ph7 been missing this feature lately, happy to merge finally!

cr7pt0gr4ph7 commented:
@ronso0 Thanks for merging! Sorry for not being involved in the project for the last few weeks, had a lot other stuff to do. I'm still planning on getting my other PRs ready for merging :+1:

ronso0 commented:
Unfortunately the global hotkey removed the Ctrl+F shortcut in the Developer Tools dialog.
Fixup is #14761

ronso0 commented:
Apparently `2.6` has been created while this was targeting `main` when main was still 2.6-alpha?
At least it's not in 2.6 : \

Shall we backport?
